### PR TITLE
Exclude Archived Cards from Time In Lists

### LINKED ISSUE #39: Time Spent in Lists does not account for archived cards
## Description
The Time Spent in Lists feature utilizes the [`GET /1/boards/[board_id]/actions`](https://developers.trello.com/advanced-reference/board#get-1-boards-board-id-actions) endpoint to retrieve actions for the board. This does _not_, however, exclude cards that are archived.

The problem with that is if any individual card is archived in a list that is inclusive in a list that is within your Start / End of work, it can artificially inflate the numbers since it will grow in perpetuity.

## Proposed Solution
Ignore the [`closed`](https://github.com/jeremytregunna/ruby-trello/blob/master/lib/trello/card.rb#L14) cards when calculating the time spent in lists.

### PR DESCRIPTION
**Pull request by leviwilson**
## Description
This PR fixes #39 by not including cards that are archived. Additionally, rather than sending back simply the `card_id` as the payload, it sends back more information that could be leveraged to add additional functionality on the client-side as well.

### Other Things
* updated to the latest `ruby-trello`
* introduced `Rack::Test` to test controller methods
  * `rspec-its` and `factory_girl` came with it as well
* the `/api` methods were DRY'd up in their usage of `new`ing up the `Trello::Client`
* updated the label count analyzer to have a secondary sort based on the label name
  * locally I was running into issues with this as the order in CI was different than the order locally that came back; this makes the specs (as well as viewing it) more predictable when there are ties on the label counts

### CODE REVIEW

**Comment by leviwilson** [comment]
When the cukes run for this, you may need to make sure that your test board has archived cards. Before, the lists were just being closed (but I don't believe this archives the cards that were in them). I've added a mixin for archiving all cards in a list (based on [this method](https://github.com/jeremytregunna/ruby-trello/blob/1811e35c640c6918b60c8e3cee8777ced681d4ee/lib/trello/list.rb#L102) that has not been released yet).

**Comment by leviwilson** [comment]
Here is the latest passing build as well (since the PRs won't successfully run): https://travis-ci.org/leviwilson/ollert/builds/151040319

**Comment by larryprice** [comment]
Hmm a bunch of my tests are failing locally. For TimeTracker, I've got off-by-one errors on (3 -> 4, 0 -> 1, 2 -> 3), and the CFD has an extra card.

For the cukes, my test run seems to get stuck on the test with a tag of "existing lists".

Thoughts? My ruby is a little rusty, so it's taking me longer to look this over than expected.

**Comment by leviwilson** [comment]
Do you have a fork that you could run against travis to see if _that_ is good? Not sure what would be different? Any specific specs?

**Comment by larryprice** [comment]
Figured out some of the time_tracker stuff - time zones! i've got a fix i'll look into merging in soonish.

Specs that currently fail for me:

    returns default cfd data (FAILED - 1)
    returns cfd data with givens tarting and ending list (FAILED - 2)

Here's the result of running rake:spec: https://paste.ubuntu.com/22982023/

The weird thing is it seems to also be affecting me on master? I'll try to dig more later this week.

**Comment by leviwilson** [comment]
Ahh! I wonder if we should add [timecop](https://github.com/travisjeffery/timecop) for the specs to avoid that...

**Comment by larryprice** [comment]
If you get a moment, you can verify my timezone commit looks sane https://github.com/sep/ollert/pull/41

**Comment by leviwilson** [comment]
I could be missing something, but I'm wondering why I did [`.utc` here](https://github.com/leviwilson/ollert/blob/exclude_closed_cards_for_til/utils/analyzers/time_tracker.rb#L19) in the first place...when I do [`to_date` later on](https://github.com/leviwilson/ollert/blob/exclude_closed_cards_for_til/utils/analyzers/time_tracker.rb#L51-L52).

If I remove the `.utc` on line 19 they pass for me.

**Comment by leviwilson** [comment]
That's another way of saying they fail locally for me, right now...'cause timezones :)

image

File: test/spec/utils/analyzers/time_tracker_spec.rb [hunk 1]
```diff
@@ -191,7 +220,7 @@ def previous_list
     end
 
     def previous_date
-      actions.last.date.to_date
+      actions.last.date.to_datetime.utc.to_date
```
Comment (larryprice) [inline test/spec/utils/analyzers/time_tracker_spec.rb]: I believe this is redundant. Can you confirm that the to_datetime.utc is not required?

Comment (larryprice) [inline test/spec/utils/analyzers/time_tracker_spec.rb]: If we (or Trello) are in control of all these actions, they should all be UTC already.

File: utils/analyzers/time_tracker.rb [hunk 2]
```diff
@@ -47,8 +48,8 @@ def initialize
         end
 
         def add(start_time, end_time)
-          start_day = start_time.to_date
-          end_day = end_time.to_date
+          start_day = start_time.to_datetime.utc.to_date
```
Comment (larryprice) [inline utils/analyzers/time_tracker.rb]: Same here with the UTC stuff... If we (or Trello) are in control of all these actions, they should all be UTC already.

File: web.rb [hunk 3]
```diff
@@ -6,6 +6,7 @@
 require 'sinatra/base'
 require 'sinatra/respond_with'
 require 'trello'
+require 'trello/core_ext/string'
```
Comment (larryprice) [inline web.rb]: what does this gain us?

Comment (leviwilson) [inline web.rb]: In the previous version of the gem, there was an extension called `json_into` that you'd use like `json_into(Trello::Card)`. When I updated the gem, you have to require this specifically to get the same behavior.

File: routes/analysis_routes.rb [hunk 4]
```diff
@@ -9,59 +9,41 @@ class Ollert
...
      params["startingList"], params["endingList"]).to_json
     status 200
   end
 
   get '/api/v1/listchanges/:board_id' do |board_id|
-    client = Trello::Client.new developer_public_key: ENV['PUBLIC_KEY'], member_token: params['token']
-
-    lists = client.get("/boards/#{board_id}/lists", filter: 'open').json_into(Trello::List)
-    all = Utils::Fetchers::ListActionFetcher.fetch(client, board_id)
+    lists = @client.get("/boards/#{board_id}/lists", filter: 'open').json_into(Trello::List)
```
Comment (larryprice) [inline routes/analysis_routes.rb]: We might consider abstracting these two calls for `lists` and `cards`.

**Comment by larryprice** [comment]
I left a couple of small comments and then we can wrap this up.

File: utils/analyzers/time_tracker.rb (hunk 2 above)
Comment (leviwilson) [inline utils/analyzers/time_tracker.rb]: I'll check it out again, but this was required to get the specs to pass after 🕗 PM. So the setup may have been incorrect? I will revisit. Whatever it is, the data needs to be faked to whatever Trello is _actually_ giving us. It was a week ago, so I need to look at it again.

I'm pretty sure the issue was it comes in as a string with the UTC information in it, but `to_date` was not taking it into account. By doing `to_datetime` to preserve the time offset, then `.utc.to_date` you got the correct date.

File: web.rb (hunk 3 above)
Comment (leviwilson) [inline web.rb]: I'm not sure what combination is different now, but these are passing without the need to do any conversion to `utc` anywhere `¯\_(ツ)_/¯`. I removed it out of `time_tracker.rb`. As it doesn't need it now. My guess is that somewhere in the builder before (in the spec setup) it was introducing the issue, but the for realz code was fine. At any rate, please look it over; it's passing after 🕗 PM here so it must be good :)

I did verify that the date strings that come back in for real Trello look like `"2016-08-14T21:38:56-04:00"`, so it has timezone information in it (just like `DateTime.now` does).

File: routes/analysis_routes.rb (hunk 4 above)
Comment (leviwilson) [inline routes/analysis_routes.rb]: Maybe want to wait until they're re-used anywhere else?

**Comment by larryprice** [comment]
alright i merged this in - we'll make sure all the tests pass and then move on with our lives 😸

**Pull request merged**